File: app/src/main/java/de/jbamberger/irremote/MainActivity.java
package de.jbamberger.irremote;

import android.content.Context;
import android.hardware.ConsumerIrManager;
import android.os.Bundle;
import android.os.Handler;
import android.support.v7.app.AppCompatActivity;
import android.view.Menu;
import android.view.MenuItem;
import android.view.MotionEvent;
import android.view.View;
import android.widget.Button;
import android.widget.TableLayout;
import android.widget.TableRow;
import android.widget.Toast;

import java.io.IOException;
import java.io.InputStream;
import java.util.Map;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

import timber.log.Timber;


public class MainActivity extends AppCompatActivity {

    private Map<String, RemoteParser.RemoteDefinition> remotes;
    private Executor exec;
    private TableLayout remoteLayout;
    private RemoteParser.IrDef commands;
    private boolean initMenu = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        remoteLayout = findViewById(R.id.main_layout);

        exec = Executors.newSingleThreadExecutor();
        exec.execute(this::loadRemotes);

    }

    public void runIR(String command) {
        final int frequency = commands.getFrequency();
        final int[] code = commands.getCodeMap().get(command);
        exec.execute(() -> {
            ConsumerIrManager mIRManager =
                    (ConsumerIrManager) this.getSystemService(Context.CONSUMER_IR_SERVICE);
            if (mIRManager == null || !mIRManager.hasIrEmitter()) {
                Timber.d("An error occurred while transmitting.");
                return;
            }
            ConsumerIrManager.CarrierFrequencyRange[] range = mIRManager.getCarrierFrequencies();
            for (ConsumerIrManager.CarrierFrequencyRange carrierFrequencyRange : range) {
                if (carrierFrequencyRange.getMinFrequency() <= frequency
                        && frequency <= carrierFrequencyRange.getMaxFrequency()) {
                    mIRManager.transmit(frequency, code);
                    return;
                }
            }
            Timber.d("The required frequency range is not supported.");

        });
//        String command = (String) v.getTag();
//        IRSenderService.startActionSendIrCode(this, Remotes.CEIL_REMOTE, command);
    }

    private void onRemotesReady(Map<String, RemoteParser.RemoteDefinition> remotes) {
        this.remotes = remotes;
    }

    @Override
    public boolean onPrepareOptionsMenu(Menu menu) {

        if (remotes != null) {
            menu.clear();
            remotes.keySet().forEach(menu::add);
        }

        return super.onPrepareOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (remotes != null) {
            RemoteParser.RemoteDefinition definition = remotes.get(item.getTitle().toString());
            initUi(definition);
        }


        return super.onOptionsItemSelected(item);
    }

    private void onRemotesLoadingFailed() {
        Toast.makeText(this, "Could not load remotes", Toast.LENGTH_LONG).show();
    }

    private void initUi(RemoteParser.RemoteDefinition def) {
        remoteLayout.removeAllViews();
        initMenu = true;
        if (def == null) {
            return;
        }
        this.commands = def.getCommandDefs();

        for (RemoteParser.Control[] row : def.getLayout().getControls()) {
            TableRow tr = new TableRow(this);
            for (RemoteParser.Control control : row) {
                Button b = new Button(this);
                if (control == null) {
                    b.setVisibility(View.INVISIBLE);
                } else {
                    b.setOnTouchListener(new View.OnTouchListener() {
                        private Handler handler;

                        @Override public boolean onTouch(View v, MotionEvent event) {
                            switch(event.getAction()) {
                                case MotionEvent.ACTION_DOWN:
                                    if (handler != null) return true;
                                    handler = new Handler();
                                    handler.post(action);
                                    break;
                                case MotionEvent.ACTION_UP:
                                    if (handler == null) return true;
                                    handler.removeCallbacks(action);
                                    handler = null;
                                    break;
                            }
                            return false;
                        }

                        private Runnable action = new Runnable() {
                            @Override public void run() {
                                exec.execute(() -> runIR(control.getCommand()));
                                handler.postDelayed(this, 300);
                            }
                        };
                    });
                    b.setTag(control.getCommand());
                    b.setText(control.getName());
                }
                tr.addView(b);
            }
            remoteLayout.addView(tr);
        }
    }

    private void loadRemotes() {
        try (InputStream is = getResources().openRawResource(R.raw.remotes)) {
            final String s = Utils.readString(is);
            final RemoteParser p = new RemoteParser();
            final Map<String, RemoteParser.RemoteDefinition> remotes = p.parse(s);
            runOnUiThread(() -> onRemotesReady(remotes));
        } catch (IOException e) {
            e.printStackTrace();
            runOnUiThread(this::onRemotesLoadingFailed);
        }
    }
}

File: app/src/main/java/de/jbamberger/irremote/MainActivity.kt
package de.jbamberger.irremote

import android.annotation.SuppressLint
import android.content.Intent
import android.hardware.ConsumerIrManager
import android.os.*
import android.util.AttributeSet
import android.view.*
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import timber.log.Timber
import java.io.IOException
import java.util.concurrent.Executors
import java.util.function.Consumer


class MainActivity : AppCompatActivity() {

    private var remotes: Map<String, RemoteParser.RemoteDefinition>? = null
    private val exec = Executors.newSingleThreadExecutor()
    private var remoteLayout: TableLayout? = null
    private var commands: RemoteParser.IrDef? = null
    private var initMenu = false
    private var irManager: ConsumerIrManager? = null
    private lateinit var vibrator: Vibrator

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

//        startActivity(Intent(this, TestActivity::class.java))
//        return

        setContentView(R.layout.activity_main)

        remoteLayout = findViewById(R.id.main_layout)
        irManager = this.getSystemService(ConsumerIrManager::class.java)
        vibrator = this.getSystemService(Vibrator::class.java)
        val errorText = findViewById<TextView>(R.id.error_text)

        if (irManager == null || !irManager!!.hasIrEmitter()) {
            Toast.makeText(this, "This device does not support Infrared communication.",
                    Toast.LENGTH_LONG).show()
        } else {
            errorText.visibility = View.GONE
            loadRemotes()
        }

    }

    fun runIR(command: String) {
        val frequency = commands!!.frequency
        val code = commands!!.codeMap[command]
        exec.execute {
            val irManager = this.irManager ?: return@execute

            val range = irManager.carrierFrequencies
            for (carrierFrequencyRange in range) {
                if (carrierFrequencyRange.minFrequency <= frequency
                        && frequency <= carrierFrequencyRange.maxFrequency) {
                    if (vibrator.hasVibrator()) {
                        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                            vibrator.vibrate(VibrationEffect.createOneShot(
                                    50, VibrationEffect.DEFAULT_AMPLITUDE))
                        } else {
                            @Suppress("DEPRECATION")
                            vibrator.vibrate(50)
                        }
                    }
                    irManager.transmit(frequency, code)
                    return@execute
                }
            }
            Timber.d("The required frequency range is not supported.")

        }
    }

    private fun onRemotesReady(remotes: Map<String, RemoteParser.RemoteDefinition>) {
        this.remotes = remotes
    }

    override fun onPrepareOptionsMenu(menu: Menu): Boolean {
        if (remotes != null) {
            menu.clear()
            remotes!!.keys.forEach(Consumer { menu.add(it) })
        }

        return super.onPrepareOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (remotes != null) {
            val definition = remotes!![item.title.toString()]
            initUi(definition)
        }


        return super.onOptionsItemSelected(item)
    }

    private fun onRemotesLoadingFailed() {
        Toast.makeText(this, "Could not load remotes", Toast.LENGTH_LONG).show()
    }


    private fun initUi(def: RemoteParser.RemoteDefinition?) {
        remoteLayout!!.removeAllViews()
        initMenu = true
        if (def == null) {
            return
        }
        this.commands = def.commandDefs

        for (row in def.layout.controls) {
            val tr = TableRow(this)
            val params = TableLayout.LayoutParams(
                    TableLayout.LayoutParams.MATCH_PARENT, TableLayout.LayoutParams.WRAP_CONTENT, 1.0f)
            tr.layoutParams = params
            for (control in row) {
                val b = Button(this)
                val buttonParams = TableRow.LayoutParams(
                        TableRow.LayoutParams.MATCH_PARENT, TableRow.LayoutParams.MATCH_PARENT, 1.0f)
                b.layoutParams = buttonParams
                if (control == null) {
                    b.visibility = View.INVISIBLE
                } else {
                    b.setOnTouchListener(
                    object : View.OnTouchListener {
                        private var handler: Handler? = null

                        private val action = object : Runnable {
                            override fun run() {
                                exec.execute { runIR(control.command) }
                                handler!!.postDelayed(this, 300)
                            }
                        }

                        override fun onTouch(v: View, event: MotionEvent): Boolean {
                            when (event.action) {
                                MotionEvent.ACTION_DOWN -> {
                                    if (handler == null) {
                                        handler = Handler(Looper.getMainLooper())
                                        handler!!.post(action)
                                    }
                                    return true
                                }
                                MotionEvent.ACTION_UP, MotionEvent.ACTION_CANCEL -> {
                                    handler?.removeCallbacks(action)
                                    handler = null
                                    return true
                                }
                            }
                            return false
                        }
                    })
                    b.tag = control.command
                    b.text = control.name
                }
                tr.addView(b)
            }
            with(remoteLayout!!) {
                isMeasureWithLargestChildEnabled = true
                addView(tr)
            }
        }
    }

    private fun loadRemotes() = exec.execute {
        try {
            resources.openRawResource(R.raw.remotes)
                    .use {
                        val remotes = RemoteParser().parse(Utils.readString(it))
                        runOnUiThread { onRemotesReady(remotes) }
                    }
        } catch (e: IOException) {
            e.printStackTrace()
            runOnUiThread { this.onRemotesLoadingFailed() }
        }
    }
}
